src/string/mod.rs

```diff
@@ -677,6 +677,138 @@ impl<LenT: LenType, S: StringStorage + ?Sized> StringInner<LenT, S> {
     pub fn clear(&mut self) {
         self.vec.clear();
     }
+
+    /// Inserts a character into this `String` at a byte position.
+    ///
+    /// This is an *O*(*n*) operation as it requires copying every element in the
+    /// buffer.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `idx` is larger than the `String`'s length, or if it does not
+    /// lie on a [`char`] boundary.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use heapless::String;
+    ///
+    /// let mut s: String<4> = String::new();
+    ///
+    /// s.insert(0, 'f').unwrap();
+    /// s.insert(1, 'o').unwrap();
+    /// s.insert(2, 'o').unwrap();
+    ///
+    /// assert_eq!("foo", s);
+    /// # Ok::<(), heapless::CapacityError>(())
+    /// ```
+    #[inline]
+    pub fn insert(&mut self, idx: usize, ch: char) -> Result<(), CapacityError> {
+        assert!(self.is_char_boundary(idx), "index must be a char boundary");
+
+        let len = self.len();
+        let ch_len = ch.len_utf8();
+
+        // Check if there is enough capacity
+        if len + ch_len > self.capacity() {
+            return Err(CapacityError);
+        }
+
+        // SAFETY: Move the bytes starting from `idx` to their new location `ch_len`
+        // bytes ahead. This is safe because we checked `len + ch_len` does not
+        // exceed the capacity and `idx` is a char boundary
+        unsafe {
+            core::ptr::copy(
+                self.vec.as_ptr().add(idx),
+                self.vec.as_mut_ptr().add(idx + ch_len),
+                len - idx,
+            );
+        }
+
+        // SAFETY: Copy the encoded character into the vacated region if
+        // `idx != len`, or into the uninitialized spare capacity otherwise.
+        unsafe {
+            // 4 bytes is the maximum length of a UTF-8 character
+            let mut buf = [0u8; 4];
+            let encoded = ch.encode_utf8(&mut buf);
+            core::ptr::copy_nonoverlapping(
+                encoded.as_ptr(),
+                self.vec.as_mut_ptr().add(idx),
+                ch_len,
+            );
```

Comment:
This differs from the `std::String` implementation which uses an unstable method to encode directly into the destination buffer.

```rust
core::char::encode_utf8_raw_unchecked(ch as u32, self.vec.as_mut_ptr().add(idx));
```

There's probably a way than what I've done here.

Reply:
Can't we directly pass a mutable ref of `self.vec` to `char::encode_utf8` directly? Something like:

```rust
let buf = (&mut self.vec)[idx..idx+4];
ch.encode_utf8(buf);
```

Reply:
I updated the PR with the latter, but I wasn't able to tell which was actually better in the godbolt output. Not because they're functionally identical, I'm just not good enough at reading it yet lol. I'll go with whatever you think makes more sense.